rclcpp/include/rclcpp/utilities.hpp

```diff
@@ -282,6 +313,113 @@ sub_will_underflow(const T x, const T y)
 }
 
+/// Wrapper which will check for over and underflow of subtractions
+/**
+ * Function will check, if
+ * x - y
+ * is permissible. If not, it will throw an according exception
+ *
+ * \tparam T must be arithmetic
+ * \param x lhs of the subtraction
+ * \param y rhs of the subtraction
+ * \throws std::underflow_error if subtraction underflows
+ * \throws std::overflow_error if subtraction overflows
+ */
+template<typename T>
+void
+check_sub(const T x, const T y)
+{
+  static_assert(std::is_arithmetic<T>::value, "must be arithmetic");
+
+  if (sub_will_overflow(x, y)) {
+    const auto error_message = std::string("subtraction leads to overflow of ") +
+      std::string(typeid(x).name());
+    throw std::overflow_error(error_message);
+  }
+  if (sub_will_underflow(x, y)) {
+    const auto error_message = std::string("subtraction leads to underflow of ") +
+      std::string(typeid(x).name());
+    throw std::underflow_error(error_message);
+  }
+}
+
+/// Checks if a multiplication will lead to an overflow
+/**
+ * Function checks if
+ * res = x * y
+ * will lead to an overflow res, where res has the data type of x
+ *
+ * \tparam X data type of the left hand side
+ * \tparam Y data type of the right hand side
+ * \param x value of the left hand side
+ * \param y value of the right hand side
+ * \return true, if multiplication overflows
+ */
+template<typename X, typename Y>
+bool
+mult_will_overflow(X x, Y y) noexcept
+{
+  if (x < 0) {
+    return (y < 0) && (x < std::numeric_limits<X>::max() / y);
+  } else {
+    return (y > 0) && (x > std::numeric_limits<X>::max() / y);
+  }
+}
+
+/// Checks if a multiplication will lead to an underflow
+/**
+ * Function checks if
+ * res = x * y
+ * will lead to an underflow of res, where res has the data type of x
+ *
+ * \tparam X data type of the left hand side
+ * \tparam Y data type of the right hand side
+ * \param x value of the left hand side
+ * \param y value of the right hand side
+ * \return true, if multiplication underflows
+ */
+template<typename X, typename Y>
+bool
+mult_will_underflow(X x, Y y) noexcept
+{
+  if (x < 0) {
+    return (y > 0) && (x < std::numeric_limits<X>::min() / y);
+  } else {
+    return (y < 0) && (x > std::numeric_limits<X>::min() / y);
+  }
+}
+
+/// Wrapper throws if a multiplication either over or underflows
+/**
+ * Function checks if
+ * res = x * y
+ * will lead to an over or underflow of res, where res has the data type of x
+ *
+ * \tparam X data type of the left hand side, must be arithmetic
+ * \tparam Y data type of the right hand side, must be arithmetic
+ * \param x value of the left hand side
+ * \param y value of the right hand side
+ * \throws std::overflow_error if multiplication overflows
+ * \throws std::underflow_error is multiplication underflows
+ */
+template<typename X, typename Y>
+void check_mult(X x, Y y)
+{
+  static_assert(std::is_arithmetic<X>::value, "X must be arithmetic");
```

Comment:
I wonder if these checks should be in the `check_*` functions or rather in the actual `*_will_overflow` functions since both are public functions and only the `*_will_overflow` functions operate on the template types.

Reply:
agree